Ending sequence for level 1 and transition to level 2 (also includes time functionality + hurry music)

Seafoamm opened the pull request:
Original Task: Flag, fireworks, end sequence, transition to next level

## Features

> Did you fully complete the assigned features?
Yes, except for the fireworks.

> What features did you Complete?
- Added in BigCastle as background asset
- Added in Flag as asset (flag goes down when mario touches it)
- Added in ability to disable input in game engine.
- Added in flag, castle, bush, block, music, labels, into levels
- Added in time functionality into SceneManager as well as hurry music when time hits 99
- Added in pulling flag state to mario, along with the animations
- Added in special scripting updates to mario when in end sequence: this includes sliding down the flag, scoring depending on where he touches, flipping and jumping off and slow walking to the castle
- Timer counts down to 0 at end of level and adds to score proportionally
- Level 1-1 transitions to 1-2 and spawns mario in to 1-2

> What features were you supposed to complete but didn't? (If any)
- The fireworks that appear when you complete the level with timer ending in 1, 3, or 6

> What features did you complete beyond your assigned features? (If any)
- All features mentioned in above list (including unassigned features)

> How did you go about implementing your features?
- Looking at the code structure and design already implemented and trying to fit within the model. Only changing places we needed to and only adding extra features when we needed to.


## Bugs

> Did you introduce any bugs or potential ones?
The hurry music for 1-2 is not playing correctly when the timer hits for some reason. There is a problem with not being able to read it from the cache. 

If possible please put file and line numbers to aid in the debugging process.
- scenemanager.js lines 180-183 (pause current music and play hurry music)
- levels.js line 188 (path to hurry music)
- scenemanager.js lines 142-143 (was already there but could conflict with our added code)

## Contributors to the Feature
- Andrew Treece
- Brandon Lu

## Final check

Have you filled out this document and gave the pull request a descriptive name (Up on top of this document and the GitHub PR title)? Answer *Yes* and that means you are good to go! Thank you for your contribution.

- [X] Yes
- [ ] No

kvietcong commented:
**Bug:** When going to 1-2 and using the pipe to get back into 1-1, sliding down the flag again drops you off the left side and proceeds to soft-brick the game. This might be an issue with the level loader though. (Same if you die and respawn)

kvietcong commented on gameengine.js:
```diff
@@ -114,9 +111,41 @@ class GameEngine {
+        this.ctx.canvas.addEventListener("keyup", that.keyup, false);
     };
 
+    disableInput() {
```
Maybe move to a flag based system rather than a function manipulating listeners? Like have a `isInputEnabled` property on the game engine and check that property in the input functions above (starting line 46). This way you never have to worry about duplicate listeners or making sure you called startInput again. It reduces the amount of states you have to think about. Optional recommendation.

kvietcong commented on sceneManager.js:
```diff
@@ -162,6 +177,22 @@ class SceneManager {
 
     update() {
         this.menuButtonTimer += this.game.clockTick;
+        if (this.time === 99) {
```
Instead of manipulating your own custom timer, I would just use `this.game.clockTick` directly. `this.time` -> `this.timeLeft` for clarity?

```js
update() {
  // ...
  if (!this.title && !this.transition && !this.paused) {
    this.timeLeft -= this.game.clockTick;
    if (this.timeLeft.round() == 99) {
      // Play Music Logic
    }
  }
}
```

Now the code above is pseudo-ish. There is a change you double play (the starting plays two times), but I find it unlikely and not too disturbing as the double play with be within frames. If you want to be a bit more rigorous, you can have a flag to make sure it doesn't double play. I think this will fix your problem. Making the logic a bit more simple and having less state to keep track of (no need for `this.timer` anymore)

kvietcong reviewed (commented):
Nice work so far. Bugs withstanding, this is a great effort. I would urge you guys to look at my comments and please do find a way to remove your personal IDE configuration files (essentially everything in `.idea` and add it to the gitignore).

kvietcong reviewed (approved):
Looks good. I see the change you made and thanks for tidying up some of the areas of the code.

Thanks for the contribution

Seafoamm commented:
> 

We were going to wait till next week to leave comments about a couple of things, but I guess since you are looking at the code now, I will comment now. 
1. We changed the timer for time left in the level like you suggested, but we remembered why we had the old time system in place with a separate timer. We were trying to emulate the behavior of the online version of mario that Chris had us reference. If we simply subtract the clock tick, it is too slow. One solution could be subtracting a multiple of the clock tick (like 3 * clocktick). Do you think that is a good idea?
2.  There is a bug where mario seems to be added in first sometimes. This seems to happen when we load into a bonus level (from a tube) and when we come back out from the tube. 
3. Currently we initiate timeLeft when we load a level. There might need to be a change of placement of that line in the future if we want time to carry over when we are entering and exiting tubes.
4. The tube in 1-2 should not go back to levelOne so the flag bug should not have been encounterable. However, we fixed the flag bug anyways in case there is a tube that is supposed to go back to level one.

Seafoamm commented:
> 

Sorry, to clarify my statement about mario being added in first, I meant Mario being added in first to the game engine's entity list. This causes him to draw behind all entities that you can walk through, like bushes (it doesn't matter with entities you collide with because you can't go behind them anyways).

kvietcong commented:
> Sorry, to clarify my statement about mario being added in first, I meant Mario being added in first to the game engine's entity list. This causes him to draw behind all entities that you can walk through, like bushes (it doesn't matter with entities you collide with because you can't go behind them anyways).

I believe that his can be fixed in the `loadLevel` function where you guys put that check for Mario. Instead of that checking you can get the index of Mario, pop him out if he exists (or create a new instance as you do), and then add him back in.

kvietcong commented:
> > 
> 
> We were going to wait till next week to leave comments about a couple of things, but I guess since you are looking at the code now, I will comment now.
> 
> 1. We changed the timer for time left in the level like you suggested, but we remembered why we had the old time system in place with a separate timer. We were trying to emulate the behavior of the online version of mario that Chris had us reference. If we simply subtract the clock tick, it is too slow. One solution could be subtracting a multiple of the clock tick (like 3 * clocktick). Do you think that is a good idea?
> 2. There is a bug where mario seems to be added in first sometimes. This seems to happen when we load into a bonus level (from a tube) and when we come back out from the tube.
> 3. Currently we initiate timeLeft when we load a level. There might need to be a change of placement of that line in the future if we want time to carry over when we are entering and exiting tubes.
> 4. The tube in 1-2 should not go back to levelOne so the flag bug should not have been encounterable. However, we fixed the flag bug anyways in case there is a tube that is supposed to go back to level one.

1. That's weird. This means that on of the implementations is running 3x slower or one 3x faster? I would prefer not to have a magical * 3 in the code but if you guys think it makes the game feel better, feel free to do so
2. Above
3. Since Mario is already a monolith-class, it might be best to keep it there. This way, mario decides when he is done with a level. Or you can open up some way for the scene manager to keep track of this.
4. Nice Work!
